File: LuaJEngine/src/main/java/org/anime_game_servers/luaj_engine/LuaJTable.java
package org.anime_game_servers.luaj_engine;

import lombok.val;
import org.anime_game_servers.lua.engine.LuaTable;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Set;

public class LuaJTable implements LuaTable {
    private final org.luaj.vm2.LuaTable table;
    private final LuaJEngine engine;

    LuaJTable(LuaJEngine engine, org.luaj.vm2.LuaTable table) {
        this.table = table;
        this.engine = engine;
    }

    @Override
    public boolean has(String key) {
        return !table.get(key).isnil();
    }

    @Override
    public Object get(String key) {
        return new LuaJValue(engine, table.get(key));
    }

    @Override
    public Object get(int key) {
        return new LuaJValue(engine, table.get(key));
    }

    @Override
    public int getInt(String key) {
        return table.get(key).checkint();
    }

    @Override
    public int optInt(String key, int defaultValue) {
        return table.get(key).optint(defaultValue);
    }

    @Override
    public int getInt(int key) {
        return table.get(key).checkint();
    }

    @Override
    public int optInt(int key, int defaultValue) {
        return table.get(key).optint(defaultValue);
    }

    @Override
    public void set(int key, int value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(String key, int value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public double getDouble(String key) {
        return table.get(key).checkdouble();
    }

    @Override
    public double optDouble(String key, double defaultValue) {
        return table.get(key).optdouble(defaultValue);
    }

    @Override
    public double getDouble(int key) {
        return table.get(key).checkdouble();
    }

    @Override
    public double optDouble(int key, double defaultValue) {
        return table.get(key).optdouble(defaultValue);
    }

    @Override
    public float getFloat(String key) {
        return table.get(key).tofloat();
    }

    @Override
    public float optFloat(String key, float defaultValue) {
        return (float) table.get(key).optdouble(defaultValue);
    }

    @Override
    public float getFloat(int key) {
        return table.get(key).tofloat();
    }

    @Override
    public float optFloat(int key, float defaultValue) {
        return (float) table.get(key).optdouble(defaultValue);
    }

    @Override
    public void set(int key, double value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(String key, double value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(int index, float value) {
        table.set(index, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(@NotNull String key, float value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public boolean getBoolean(String key) {
        return table.get(key).toboolean();
    }

    @Override
    public boolean optBoolean(String key, boolean defaultValue) {
        return table.get(key).optboolean(defaultValue);
    }

    @Override
    public boolean getBoolean(int key) {
        return table.get(key).toboolean();
    }

    @Override
    public boolean optBoolean(int key, boolean defaultValue) {
        return table.get(key).optboolean(defaultValue);
    }

    @Override
    public void set(int key, boolean value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(String key, boolean value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public String getString(String key) {
        return table.get(key).tojstring();
    }

    @Override
    public String optString(String key, String defaultValue) {
        return table.get(key).optjstring(defaultValue);
    }

    @Override
    public String getString(int key) {
        return table.get(key).toString();
    }

    @Override
    public String optString(int key, String defaultValue) {
        return table.get(key).optjstring(defaultValue);
    }

    @Override
    public void set(int key, String value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public void set(String key, String value) {
        table.set(key, org.luaj.vm2.LuaValue.valueOf(value));
    }

    @Override
    public LuaTable getTable(String key) {
        val variable = table.get(key);
        if (!variable.istable()) {
            return null;
        }
        return new LuaJTable(engine, variable.checktable());
    }

    @Override
    public LuaTable getTable(int key) {
        val variable = table.get(key);
        if (!variable.istable()) {
            return null;
        }
        return new LuaJTable(engine, variable.checktable());
    }

    @Override
    public void set(int key, LuaTable value) {
        val rawTable = value.getRawTable();
        if (rawTable instanceof org.luaj.vm2.LuaTable)
            table.set(key, (org.luaj.vm2.LuaTable) rawTable);
    }

    @Override
    public void set(String key, LuaTable value) {
        val rawTable = value.getRawTable();
        if (rawTable instanceof org.luaj.vm2.LuaTable)
            table.set(key, (org.luaj.vm2.LuaTable) rawTable);
    }

    @Override
    public Set<String> getKeys() {
        return Arrays.stream(table.keys()).map(org.luaj.vm2.LuaValue::tojstring).collect(java.util.stream.Collectors.toSet());
    }

    @Override
    public int[] getAsIntArray() {
        val array = new int[table.length()];
        for (int i = 0; i < array.length; i++) {
            array[i] = table.get(i + 1).optint(0);
        }
        return array;
    }

    @Override
    public int getSize() {
        return table.length();
    }

    @Override
    public Object getRawTable() {
        return table;
    }
}

File: LuaJEngine/src/main/java/org/anime_game_servers/luaj_engine/LuaJTable.kt
package org.anime_game_servers.luaj_engine

import org.anime_game_servers.lua.engine.LuaTable
import org.luaj.vm2.LuaValue

class LuaJTable internal constructor(private val engine: LuaJEngine, private val table: org.luaj.vm2.LuaTable) :
    LuaTable {
    override fun has(key: String): Boolean {
        return !table[key].isnil()
    }

    override fun get(key: String): Any {
        return LuaJValue(engine, table[key])
    }

    override fun get(key: Int): Any {
        return LuaJValue(engine, table[key])
    }

    override fun getInt(key: String): Int {
        return table[key].checkint()
    }

    override fun optInt(key: String, defaultValue: Int): Int {
        return table[key].optint(defaultValue)
    }

    override fun getInt(key: Int): Int {
        return table[key].checkint()
    }

    override fun optInt(key: Int, defaultValue: Int): Int {
        return table[key].optint(defaultValue)
    }

    override fun set(key: Int, value: Int) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun set(key: String, value: Int) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun getLong(key: String): Long {
        return table[key].checklong()
    }

    override fun optLong(key: String, defaultValue: Long): Long {
        return table[key].optlong(defaultValue)
    }

    override fun getLong(key: Int): Long {
        return table[key].checklong()
    }

    override fun optLong(key: Int, defaultValue: Long): Long {
        return table[key].optlong(defaultValue)
    }

    override fun set(key: Int, value: Long) {
        table[key] = LuaValue.valueOf(value.toDouble())
    }

    override fun set(key: String, value: Long) {
        table[key] = LuaValue.valueOf(value.toDouble())
    }

    override fun getDouble(key: String): Double {
        return table[key].checkdouble()
    }

    override fun optDouble(key: String, defaultValue: Double): Double {
        return table[key].optdouble(defaultValue)
    }

    override fun getDouble(key: Int): Double {
        return table[key].checkdouble()
    }

    override fun optDouble(key: Int, defaultValue: Double): Double {
        return table[key].optdouble(defaultValue)
    }

    override fun getFloat(key: String): Float {
        return table[key].tofloat()
    }

    override fun optFloat(key: String, defaultValue: Float): Float {
        return table[key].optdouble(defaultValue.toDouble()).toFloat()
    }

    override fun getFloat(key: Int): Float {
        return table[key].tofloat()
    }

    override fun optFloat(key: Int, defaultValue: Float): Float {
        return table[key].optdouble(defaultValue.toDouble()).toFloat()
    }

    override fun set(key: Int, value: Double) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun set(key: String, value: Double) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun set(key: Int, value: Float) {
        table[key] = LuaValue.valueOf(value.toDouble())
    }

    override fun set(key: String, value: Float) {
        table[key] = LuaValue.valueOf(value.toDouble())
    }

    override fun getBoolean(key: String): Boolean {
        return table[key].toboolean()
    }

    override fun optBoolean(key: String, defaultValue: Boolean): Boolean {
        return table[key].optboolean(defaultValue)
    }

    override fun getBoolean(key: Int): Boolean {
        return table[key].toboolean()
    }

    override fun optBoolean(key: Int, defaultValue: Boolean): Boolean {
        return table[key].optboolean(defaultValue)
    }

    override fun set(key: Int, value: Boolean) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun set(key: String, value: Boolean) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun getString(key: String): String? {
        return table[key].tojstring()
    }

    override fun optString(key: String, defaultValue: String?): String? {
        return table[key].optjstring(defaultValue)
    }

    override fun getString(key: Int): String {
        return table[key].toString()
    }

    override fun optString(key: Int, defaultValue: String?): String? {
        return table[key].optjstring(defaultValue)
    }

    override fun set(key: Int, value: String) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun set(key: String, value: String) {
        table[key] = LuaValue.valueOf(value)
    }

    override fun getTable(key: String): LuaTable? {
        val variable = table[key]
        if (!variable.istable()) {
            return null
        }
        return LuaJTable(engine, variable.checktable())
    }

    override fun getTable(key: Int): LuaTable? {
        val variable = table[key]
        if (!variable.istable()) {
            return null
        }
        return LuaJTable(engine, variable.checktable())
    }

    override fun set(key: Int, value: LuaTable) {
        val rawTable: Any = value.getRawTable()
        if (rawTable is org.luaj.vm2.LuaTable) table[key] =  rawTable
    }

    override fun set(key: String, value: LuaTable) {
        val rawTable: Any = value.getRawTable()
        if (rawTable is org.luaj.vm2.LuaTable) table[key] = rawTable
    }

    override fun getKeys(): Set<String> {
        return table.keys().map { it.tojstring() }.toSet()
    }

    override fun getAsIntArray(): IntArray {
        val array = IntArray(table.length())
        for (i in array.indices) {
            array[i] = table[i + 1].optint(0)
        }
        return array
    }

    override fun getAsFloatArray(): FloatArray {
        val array = FloatArray(table.length())
        for (i in array.indices) {
            array[i] = table[i + 1].optdouble(0.0).toFloat()
        }
        return array
    }

    override fun getAsDoubleArray(): DoubleArray {
        val array = DoubleArray(table.length())
        for (i in array.indices) {
            array[i] = table[i + 1].optdouble(0.0)
        }
        return array
    }

    override fun getAsStringArray(): Array<String> {
        val result = mutableListOf<String>()
        for (i in 0 until table.length()) {
            result.add(i, table[i + 1].optjstring(""))
        }
        return result.toTypedArray()
    }
    
    override fun <T> asObject(type: Class<T>): T? {
        return engine.serializer.toObject(type, table)
    }

    override fun getSize(): Int {
        return table.length()
    }

    override fun getRawTable(): Any {
        return table
    }
}
